src/include/platform/PersistedStorage.h

```diff
@@ -87,6 +91,147 @@ CHIP_ERROR Read(Key aKey, uint32_t & aValue);
  */
 CHIP_ERROR Write(Key aKey, uint32_t aValue);
 
+class KeyValueStore
+{
+public:
+    virtual ~KeyValueStore() = default;
+    /**
+     * @brief
+     *   Reads the value of an entry in the KVS. The value is read into the
+     *   provided buffer and the number of bytes read is returned. If desired,
+     *   the read can be started at an offset.
+     *
+     *   If the output buffer is too small for the value, Get returns
+     *   CHIP_ERROR_BUFFER_TOO_SMALL with the number of bytes read returned in
+     *   read_bytes_size, which should be the buffer_size.
+     *
+     *   The remainder of the  value can be read by calling get with an offset.
+     *
+     * @param[in]     key               The name of the key to get, this is a
+     *                                  NULL terminated string.
+     * @param[in,out] buffer            A buffer to read the value into.
+     * @param[in]     buffer_size       The size of the buffer in bytes.
+     * @param[in,out] read_bytes_size   The number of bytes which were
+     *                                  copied into the buffer. Optionally can
+     *                                  provide nullptr if not needed.
+     * @param[in]     offset_bytes      The offset byte index to start the read.
+     *
+     * @return CHIP_NO_ERROR the entry was successfully read
+     *         CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND the key is not
+     *                                                       present in the KVS
+     *         CHIP_ERROR_INTEGRITY_CHECK_FAILED found the entry, but the data
+     *                                           was corrupted
+     *         CHIP_ERROR_BUFFER_TOO_SMALL the buffer could not fit the entire
+     *                                     value, but as many bytes as possible
+     *                                     were written to it
+     *         CHIP_ERROR_WELL_UNINITIALIZED the KVS is not initialized
+     *         CHIP_ERROR_INVALID_ARGUMENT key is empty or too long or value is
+     *                                     too large
+     */
+    virtual CHIP_ERROR Get(const char * key, void * buffer, size_t buffer_size, size_t * read_bytes_size = nullptr,
```

Comment:
why not re-use the Key type already defined for this interface, which allows strings depending on platform configuration.  this choice forces all platforms (including those that have, e.g. uint32 keys) to support strings as keys

Reply:
The KVS is intended to work on a single device and not be used as shared storage between multiple devices, so the endianness should be the same on the write and read, and the KVS will preserve the byte order/endianness.

Reply:
@rgoliver I'd like to see `const char * key` changed to `Key key` or be offered a justification for why this won't work

Reply:
Sorry, for some reason my responses got mixed up, that was for a different comment.

Using the Key type requires using a named key for everything, and likely having a single fixed set of keys, which works if the KVS is restricted to config options but greatly reduced the flexibility for using it for anything else. This seems like a heavy cost, for not much gain. Pigweed (already in the repo) contains a generalized well tested KVS implementation which would work with string based keys and can be quickly brought up on any platforms which need it. I have already brought this up and tested it on EFR32 with good results.

Reply:
I've filed an issue to track so this isn't a blocker